core/src/test/java/com/linecorp/armeria/server/ServerListenerTest.java

```diff
@@ -0,0 +1,112 @@
+import com.linecorp.armeria.testing.server.ServerRule;
+
+public class ServerListenerTest {
+    private static long STARTING_AT = 0L;
+    private static long STARTED_AT = 0L;
+    private static long STOPPING_AT = 0L;
+    private static long STOPPED_AT = 0L;
+
+    @ClassRule
+    public static final ServerRule server = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.meterRegistry(PrometheusMeterRegistries.newRegistry());
+
+            final Service<HttpRequest, HttpResponse> immediateResponseOnIoThread =
+                    new EchoService().decorate(LoggingService.newDecorator());
+
+            sb.service("/", immediateResponseOnIoThread);
+
+            // Record when the method triggered
+            ServerListener sl = new ServerListenerBuilder()
+                    .onStarting(() -> {
+                        ServerListenerTest.STARTING_AT = System.currentTimeMillis();
+                    })
+                    .onStarted(() -> {
+                        ServerListenerTest.STARTED_AT = System.currentTimeMillis();
+                    })
+                    .onStopping(() -> {
+                        ServerListenerTest.STOPPING_AT = System.currentTimeMillis();
+                    }, () -> {
+                        ServerListenerTest.STARTING_AT = 0L;
+                    })
+                    .onStopped(() -> {
+                        ServerListenerTest.STOPPED_AT = System.currentTimeMillis();
+                    }, () -> {
+                        ServerListenerTest.STARTED_AT = 0L;
+                    })
+                    .build();
+            sb.serverListener(sl);
+        }
+    };
+
+    @Before
+    public void startServer() {
+        server.start();
+    }
+
+    @Test
+    public void testServerListener() throws Exception {
+        // Before stop
+        assertThat(ServerListenerTest.STARTING_AT).isGreaterThan(0L);
+        assertThat(ServerListenerTest.STARTED_AT).isGreaterThan(ServerListenerTest.STARTING_AT);
+        assertThat(ServerListenerTest.STOPPING_AT).isEqualTo(0L);
+        assertThat(ServerListenerTest.STOPPED_AT).isEqualTo(0L);
+
+        final Server server = ServerListenerTest.server.server();
+        server.stop().get();
+
+        // After stop
+        assertThat(ServerListenerTest.STARTING_AT).isEqualTo(0L);
+        assertThat(ServerListenerTest.STARTED_AT).isEqualTo(0L);
+        assertThat(ServerListenerTest.STOPPING_AT).isGreaterThan(0L);
+        assertThat(ServerListenerTest.STOPPED_AT).isGreaterThan(ServerListenerTest.STOPPING_AT);
+    }
+
+    private static class EchoService extends AbstractHttpService {
+        @Override
+        protected final HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req) {
+            return HttpResponse.from(req.aggregate()
+                                        .thenApply(this::echo)
+                                        .exceptionally(CompletionActions::log));
```

Comment:
`exceptionally` will trigger NPE from `HttpResponse.from()` because `CompletionActions::log` will translate an exception into a `null` value.

Given this service will never be invoked, how about just replacing this with a hello world service for even more simplicity?